### PR TITLE
Assure thread affinity

### LINKED ISSUE #89: parallel="classes" is not forcing test methods from the same testClass to be run in the same thread as it is suposed to (I guess)
Extract from documentation:

```
parallel="classes": TestNG will run all the methods in the same class in the same 
thread, but each class will be run in a separate thread. 
```

As this is not fully specified what the author means by <pre>all the methods in the same class</pre> in case when we have a @Test method in an abstract parent class and a @Test method in child class (They are in the same testClass but in different realClasses - as it is called in the source). 

I guess they should be run sequentially. 
If I am wrong there is no issue here and please close it. If I am right here is short example exposing the problem:

```
public abstract class A {
    @Test
    public void someMethodA1() throws InterruptedException {
        scenario("someMethodA1");
    }

    protected void scenario(String s) throws InterruptedException {
        System.out.println("START["+Thread.currentThread()+"]: " + s);
        synchronized (s) { 
            s.wait(1000);
        }
        System.out.println("END["+Thread.currentThread()+"]: " + s);
    }
}

public abstract class B extends A{
    @Test
    public void someMethodB1 () throws InterruptedException {
        scenario("something else");
    }
}

public class C extends B{
    //@Test
    public void veryImportant() {}
}
```

---

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="Some Suite" >  
  <test name="Some tests" parallel="classes" thread-count="3">
    <packages>
      <package name=""/>
    </packages>
  </test>
</suite>
```

If you run this suite someMethodA1 and someMethodB1 will run in parallel (as can be seen on System.out). 
If you uncomment in class C anotation @Test  someMethodA1(), someMethodB1() and veryImportant() will execute sequentially.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #89, #1050, #1066, #1173, #1185

This PR aims at assuring that methods that fall
under the below two use cases, all run on the 
same thread when classes are being run in parallel:

* @Test methods in a class are ordered by priority
* @Test methods have a single dependency on 
another method using “dependsOnMethods” attribute.

The thread affinity feature is supposed to be 
“experimental” and it can be turned on via the
JVM argument : `-Dtestng.thread.affinity=true`

This feature is turned off by default just to 
ensure that we don’t have any users experiencing
un-usual behavior.

Fixes #89, #1050, #1066, #1173, #1185

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Comment by krmahadevan** [comment]
@juherr @cbeust - I have tried to implement the approach that I called out [here](https://github.com/cbeust/testng/issues/1773#issuecomment-385856857)

**Comment by krmahadevan** [comment]
@juherr @cbeust - Any idea why Travis CI build doesn't show up on this PR ?

**Review by juherr — APPROVED** [review]
LGTM if all tests pass

**Comment by krmahadevan** [comment]
Here's the build results for this branch from my fork : https://travis-ci.org/krmahadevan/testng/builds/378773133

**Comment by dr29bart** [comment]
The PR worth a minor release at least

**Comment by krmahadevan** [comment]
@dr29bart - Can you please quickly check if the fix works using TestNG `7.0.0-SNAPSHOT` ?
Please remember to use the JVM argument `-Dtestng.thread.affinity=true`

**Comment by ryudice** [comment]
hi, Is there someway I can test this without having to compile testng myself and without waiting for a new release? Like a development maven feed

**Comment by krmahadevan** [comment]
@cbeust - Snapshots aren't being updated to maven snapshot repository. Can you please help take a look ? https://oss.sonatype.org/content/repositories/snapshots/org/testng/testng/7.0.0-SNAPSHOT/

**Comment by robinbobin25** [comment]
Hello! Is there any chance it will be released soon? 
Thanks

**Comment by juherr** [comment]
@robinbobin25 Not planned yet but you can follow https://github.com/cbeust/testng/issues/1862

**Comment by balrajHostAnalytics** [comment]
@krmahadevan is -Dtestng.thread.affinity=true fix also covers testng "timeout" with single thread.

**Comment by krmahadevan** [comment]
Nope. I dont think its related to "timeout". Its mainly to force dependent
methods to run on the same thread as their upstream methods.
But if you are asking what happens when I add a "timeout" attribute to an
"@Test" method that depends on an upstream method, run with parallelism
enabled and with the JVM argument set as "-Dtestng.thread.affinity=true",
that I am not sure. I believe the "timeout" is always a special case,
because TestNG always runs methods that have "timeout" attributes in a
separate thread instead of the current one. So this JVM flag may not hold
true for these sort of methods. But no harm in quickly trying out. Can you
please help try it out quickly and post back your findings ?

**Comment by jacksonwen001** [comment]
any updates ?

**Comment by krmahadevan** [comment]
@jacksonwen001 - What updates are you expecting? The latest released version of TestNG is `7.4.0`. Can you please be explicit in terms of what is NOT working (if you are having a problem with this feature) ?

**Pull request merged**
